src/event_listener.rs

```diff
@@ -0,0 +1,194 @@
+/// be called when a specific RocksDB event happens such as flush. It can
+/// be used as a building block for developing custom features.
+///
+/// ## Important
+///
+/// Callback functions should not run for an extended period before they
+/// return as this will slow down RocksDB.
+///
+/// ## Threading
+///
+/// All EventListener callback will be called using the
+/// actual thread that involves in that specific event. For example, it
+/// is the RocksDB background flush thread that does the actual flush to
+/// call EventListener::OnFlushCompleted().
+///
+/// ## Locking
+///
+/// All EventListener callbacks are designed to be called without
+/// the current thread holding any DB mutex. This is to prevent potential
+/// deadlock and performance issue when using EventListener callbacks
+/// in a complex way.
+pub trait EventListener {
+    /// Name identifying the listener
+    fn name(&self) -> &CStr;
+
+    /// A callback function to RocksDB which will be called whenever a
+    /// registered RocksDB flushes a file. The default implementation is
+    /// no-op.
+    ///
+    /// Note that the this function must be implemented in a way such that
+    /// it should not run for an extended period of time before the function
+    /// returns.  Otherwise, RocksDB may be blocked.
+    fn on_flush_completed(&self, _flush_job_info: &FlushJobInfo) {}
+}
+
+/// Flush information
+#[derive(Debug)]
+pub struct FlushJobInfo {
+    /// The name of the column family
+    pub cf_name: String,
+    /// The path to the newly created file
+    pub file_path: String,
+    /// The smallest sequence number in the newly created file
+    pub smallest_seqno: u64,
+    /// The largest sequence number in the newly created file
+    pub largest_seqno: u64,
+    /// Flush reason
+    pub flush_reason: FlushReason,
+}
+
+/// Flush reason
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[repr(i32)]
+pub enum FlushReason {
+    Others = 0,
+    GetLiveFiles = 1,
+    ShutDown = 2,
+    ExternalFileIngestion = 3,
+    ManualCompaction = 4,
+    WriteBufferManager = 5,
+    WriteBufferFull = 6,
+    Test = 7,
+    DeleteFiles = 8,
+    AutoCompaction = 9,
+    ManualFlush = 10,
+    ErrorRecovery = 11,
+    ErrorRecoveryRetryFlush = 12,
+    WalFull = 13,
+    CatchUpAfterErrorRecovery = 14,
+}
+
+struct EventListenerCallback<T>
+where
+    T: EventListener + 'static + Send + Sync,
+{
+    inner: Arc<T>,
+}
+
+impl<T> EventListenerCallback<T>
+where
+    T: EventListener + 'static + Send + Sync,
+{
+    pub fn new(listener: Arc<T>) -> Self {
+        Self { inner: listener }
+    }
+
+    pub unsafe extern "C" fn destructor(raw_cb: *mut c_void) {
+        drop(Box::from_raw(raw_cb as *mut Self));
+    }
+
+    pub unsafe extern "C" fn name(raw_cb: *mut c_void) -> *const libc::c_char {
+        let cb = &*(raw_cb as *mut Self);
+        cb.inner.name().as_ptr()
+    }
+
+    pub unsafe extern "C" fn on_flush_completed(
```

Comment:
Do those need to be public?

Reply:
Accidental leftover, already removed in a subsequent update!